presto-main/src/main/java/com/facebook/presto/sql/planner/plan/MultiSourceSymbolMapping.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.planner.plan;
+
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.tree.SymbolReference;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Function;
+import com.google.common.collect.FluentIterable;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ListMultimap;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
+
+import javax.annotation.concurrent.Immutable;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Function.identity;
+
+@Immutable
+public class MultiSourceSymbolMapping
+{
+    private final ListMultimap<Symbol, Symbol> outputToInputs;
+    private final List<Symbol> outputs;
+    private final List<PlanNode> sources;
+
+    public MultiSourceSymbolMapping(ListMultimap<Symbol, Symbol> outputToInputs, List<PlanNode> sources)
+    {
+        this(outputToInputs, ImmutableList.copyOf(requireNonNull(outputToInputs, "outputToInputs is null").keySet()), sources);
+    }
+
+    public MultiSourceSymbolMapping(
+            @JsonProperty("outputToInputs") ListMultimap<Symbol, Symbol> outputToInputs,
+            @JsonProperty("outputs") List<Symbol> outputs,
+            @JsonProperty("sources") List<PlanNode> sources)
+    {
+        this.outputToInputs = ImmutableListMultimap.copyOf(requireNonNull(outputToInputs, "outputToInputs is null"));
+        this.outputs = ImmutableList.copyOf(requireNonNull(outputs, "outputs is null"));
+        this.sources = ImmutableList.copyOf(requireNonNull(sources, "sources is null"));
+
+        checkArgument(!sources.isEmpty(), "Must have at least one source");
```

Comment:
make sure that for each output symbol there is the same number of input symbols (see `ExchangeNode` constructor)